sklearn/naive_bayes.py

```diff
@@ -958,3 +985,245 @@ def _joint_log_likelihood(self, X):
+        ----------
+        X : {array-like, sparse matrix}, shape = [n_samples, n_features]
+            Training vectors, where n_samples is the number of samples and
+            n_features is the number of features. Here, each feature of X is
+            assumed to be from a different categorical distribution.
+
+        y : array-like, shape = [n_samples]
+            Target values.
+
+        sample_weight : array-like, shape = [n_samples], (default=None)
+            Weights applied to individual samples (1. for unweighted).
+
+        Returns
+        -------
+        self : object
+        """
+        self._check_settings()
+        return super(CategoricalNB, self).fit(X, y,
+                                              sample_weight=sample_weight)
+
+    def partial_fit(self, X, y, classes=None, sample_weight=None):
+        """Incremental fit on a batch of samples.
+
+        This method is expected to be called several times consecutively
+        on different chunks of a dataset so as to implement out-of-core
+        or online learning.
+
+        This is especially useful when the whole dataset is too big to fit in
+        memory at once.
+
+        This method has some performance overhead hence it is better to call
+        partial_fit on chunks of data that are as large as possible
+        (as long as fitting in the memory budget) to hide the overhead.
+
+        Parameters
+        ----------
+        X : {array-like, sparse matrix}, shape = [n_samples, n_features]
+            Training vectors, where n_samples is the number of samples and
+            n_features is the number of features. Here, each feature of X is
+            assumed to be from a different categorical distribution.
+
+        y : array-like, shape = [n_samples]
+            Target values.
+
+        classes : array-like, shape = [n_classes] (default=None)
+            List of all the classes that can possibly appear in the y vector.
+
+            Must be provided at the first call to partial_fit, can be omitted
+            in subsequent calls.
+
+        sample_weight : array-like, shape = [n_samples], (default=None)
+            Weights applied to individual samples (1. for unweighted).
+
+        Returns
+        -------
+        self : object
+        """
+        self._check_settings()
+        return (super(CategoricalNB, self)
+                .partial_fit(X, y, classes, sample_weight=sample_weight))
+
+    def _check_X(self, X):
+        return check_array(X, accept_sparse=False)
+
+    def _check_X_y(self, X, y):
+        return check_X_y(X, y, accept_sparse=False)
+
+    def _check_settings(self):
+        if self.handle_unknown not in ('ignore', 'warn', 'raise'):
+            raise ValueError("The attribute 'handle_unknown' is '{}' and "
+                             "should be one of 'ignore', 'warn' or 'raise'"
+                             .format(self.handle_unknown))
+
+    def _init_counters(self, n_effective_classes, n_features):
```

Comment:
Okay, since in `BaseDiscreteNB` we have:
```python
    def _init_counters(self, n_effective_classes, n_features):
        self.class_count_ = np.zeros(n_effective_classes, dtype=np.float64)
        self.feature_count_ = np.zeros((n_effective_classes, n_features),
                                       dtype=np.float64)
```
and we don't need the `feature_count_` variable, we have to overwrite `_init_counters` to not allocate some unused and confusing attribute, thus:
```python 
    def _init_counters(self, n_effective_classes, n_features):
        self.class_count_ = np.zeros(n_effective_classes, dtype=np.float64)
        self.category_count_ = [np.zeros((self.class_count_.shape[0], 0))
                                for _ in range(n_features)]
        self.feature_cat_index_mapping_ = [{} for _ in range(n_features)]
```
The question is if `feature_cat_index_mapping_` should better be placed before `super(...)` in `fit`. For `partial_fit` to work we would have to make sure it's called only once, meaning we would need to `_check_partial_fit_first_call` which is currently done implicitly for us in `BaseDiscreteNB.partial_fit`. Subsequently, I am not sure if removing the initialization of `feature_cat_index_mapping_` and putting it in `fit` and `partial_fit` is really nicer than the current solution.

Reply:
I was think of something like this.


```python
def fit():
   self.feature_cat_index_mapping_ = [{} for _ in range(n_features)]
   super()....
def partial_fit():
   if _check_partial_fit_first_call():
		   self.feature_cat_index_mapping_ = ...
   super()...
```

Reply:
@NicolasHug ok, we will change it to that.

Reply:
@NicolasHug I want to add my thoughts to the discussion:
True, `feature_cat_index_mapping_` is not a counter. However, it is closely related to the counters, because it acts as a translator between the data and the counters.
Additionally, there is already a check `if _check_partial_fit_first_call(self, classes):` in `partial_fit()` of BaseDiscreteNB. We would implement the functionality twice then, at different places.

So in my opinion, it is cleaner to initiate `feature_cat_index_mapping_` in `_init_counters`. Feel free to disagree, you have the last word.

Reply:
that's just a suggestion, feel free to keep the code as-is

Reply:
Okay, I kept the code as-is then